File: backend-node/src/endpoints/dtos.ts
import { Money } from '../domain/money.js';
import { Refund, Transaction } from '../domain/transaction.js';

export interface MoneyDto {
  units: number;
  currency: string;
}

export interface RefundDto {
  id: string;
  transactionId: string;
  amount: MoneyDto;
  feeReversed: MoneyDto;
  reason: string;
  idempotencyKey: string;
  createdAt: string;
}

export interface TransactionDto {
  id: string;
  merchantId: string;
  gross: MoneyDto;
  merchantFee: MoneyDto;
  maskedPan: string;
  status: string;
  createdAt: string;
  refundableRemaining: MoneyDto;
  refunds: RefundDto[];
}

export interface ApiErrorDto {
  code: string;
  message: string;
}

export function moneyToDto(m: Money): MoneyDto {
  return { units: m.units, currency: m.currency };
}

export function refundToDto(r: Refund): RefundDto {
  return {
    id: r.id,
    transactionId: r.transactionId,
    amount: moneyToDto(r.amount),
    feeReversed: moneyToDto(r.feeReversed),
    reason: r.reason,
    idempotencyKey: r.idempotencyKey,
    createdAt: r.createdAt.toISOString()
  };
}

export function transactionToDto(t: Transaction): TransactionDto {
  return {
    id: t.id,
    merchantId: t.merchantId,
    gross: moneyToDto(t.gross),
    merchantFee: moneyToDto(t.merchantFee),
    maskedPan: t.maskedPan,
    status: t.status,
    createdAt: t.createdAt.toISOString(),
    refundableRemaining: moneyToDto(t.refundableRemaining()),
    refunds: t.refunds.map(refundToDto)
  };
}

File: backend/src/PaymentLab.Api/Endpoints/Dtos.cs
using PaymentLab.Api.Domain;

namespace PaymentLab.Api.Endpoints;

public sealed record MoneyDto(long Units, string Currency)
{
    public static MoneyDto From(Money money) => new(money.Units, money.Currency);
}

public sealed record RefundDto(
    string Id,
    string TransactionId,
    MoneyDto Amount,
    MoneyDto FeeReversed,
    string Reason,
    DateTimeOffset CreatedAt)
{
    public static RefundDto From(Refund refund) => new(
        refund.Id,
        refund.TransactionId,
        MoneyDto.From(refund.Amount),
        MoneyDto.From(refund.FeeReversed),
        refund.Reason,
        refund.CreatedAt);
}

public sealed record TransactionDto(
    string Id,
    string MerchantId,
    MoneyDto Gross,
    MoneyDto MerchantFee,
    MoneyDto TotalRefunded,
    MoneyDto RefundableRemaining,
    string Status,
    string MaskedPan,
    DateTimeOffset AuthorizedAt,
    IReadOnlyList<RefundDto> Refunds)
{
    public static TransactionDto From(Transaction transaction) => new(
        transaction.Id,
        transaction.MerchantId,
        MoneyDto.From(transaction.Gross),
        MoneyDto.From(transaction.MerchantFee),
        MoneyDto.From(transaction.TotalRefunded()),
        MoneyDto.From(transaction.RefundableRemaining()),
        transaction.Status.ToString(),
        transaction.MaskedPan,
        transaction.AuthorizedAt,
        transaction.Refunds.Select(RefundDto.From).ToList());
}

/// <summary>Body of POST /api/transactions/{id}/refunds.</summary>
public sealed record RefundRequest(string Reason);

public sealed record ErrorDto(string Code, string Message);
